mempool/src/feerate.rs

```diff
@@ -0,0 +1,54 @@
+use common::primitives::amount::Amount;
+
+lazy_static::lazy_static! {
+    pub(crate) static ref INCREMENTAL_RELAY_FEE_RATE: FeeRate = FeeRate::new(1000);
+    pub(crate) static ref INCREMENTAL_RELAY_THRESHOLD: FeeRate = FeeRate::new(500);
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct FeeRate {
+    tokens_per_kb: u128,
+}
+
+impl FeeRate {
+    pub(crate) fn new(tokens_per_kb: u128) -> Self {
+        Self { tokens_per_kb }
+    }
+
+    pub(crate) fn of_tx(fee: Amount, tx_size: usize) -> Self {
+        let tokens_per_byte = Self::div_up(fee.into(), tx_size);
+        Self {
+            tokens_per_kb: tokens_per_byte * 1000,
+        }
```

Comment:
Two issues:
1. The point of multiplying by 1000 is to represent the feerate with higher precision. Dividing by the number of bytes first and returning the rounded value makes the subsequent multiplication by 1000 rather pointless. Should be something like: `(1000 * tokens) / bytes` rather than `1000 * (tokens / bytes)`. It makes a difference because we use integer division.
2. We have to be careful about overflow. Say someone issues a token with total supply of `u128::MAX` and the transaction and pays out all of it in fees in one transaction. Such transaction can weigh realistically around 120 bytes. The feerate internal representation (1000 * (u128::MAX / 120)) overflows. We should either limit either the total supply of tokens or the maximum fee to u128::MAX / 1000.

Reply:
Doing any unchecked math operations without proper Result/Option handling is basically playing with fire. This is exactly why I didn't want to allow releasing integers from Amount... it's like it's inevitable for people to do these mistakes again, and again and again. I don't know how paranoid we have to become... maybe we should just ban using normal integers and write a generic wrapper that only allows checked ops.

Reply:
@iljakuklic the first point you made is easy enough to fix but the second is a decision I have no strong feeling about since u128 sounds to me such a huge number. @TheQuantumPhysicist  What do you think regarding Lukas's second point?

Reply:
The problem isn't specifically large or small number. But the problem is that we're developing a premissionless public software. My fear is always that some operations are directly accessible to the public and can be used to crash nodes or produce undefined behavior.

Even worse. The problem is that even if the code is checked now and ensure that it's never gonna overflow because of a prior function, my paranoia still makes me think that a future refactor might reintroduce the problem, because people typically don't look at other code when they refactor, but rather only look at the code they're changing. To solve that problem, I either write tests specifically prove that this will never happen for the function in question, or make the function self-protecting such that I don't have to worry about this.